[Feat/#3] week 2 필수과제만 했어요

angryPodo opened the pull request:
## 📌 PR 요약
대충 2주차 필수과제 했다는 뜻

🌱 작업한 내용
- 의존성 세팅
- SAA 적용
- 젯팩 네비 적용
- like Clean Architecture 

🌱 To. 리뷰어
이번에 젯팩네비와 함께 hilt를 적용했습니다.
DataStore를 이용해서 자동로그인을 구현했는데 궁금한점 있으시면 코멘트로 질문 달아주세요!
코드 리뷰가 무조건 평가만 있는게 아니라 질문도 코드 리뷰입니다ㅎㅎ 언제든 궁금한건 코리ㄱㄱ

## 📸 스크린샷

https://github.com/user-attachments/assets/7c4f86bf-92ac-4fb1-818d-0c8f4f6fcd7f

## 📮 관련 이슈
- Resolved: #3

angryPodo committed:
[MOD/#3] 세팅

angryPodo committed:
[MOD/#3] 과제 밑작업...짜증

angryPodo committed:
[MOD/#3] navigation and ui components to improve user experience and fix issues with keyboard and focus management

angryPodo committed:
add new repository and viewmodel classes for user sign in and sign up functionality

angryPodo committed:
refactor signup flow to simplify navigation and validation logic

angryPodo committed:
refactor sign-in logic to include input validation and error handling for id and password length, and character requirements

angryPodo committed:
1주차 과제 리펙토링 완료

angryPodo committed:
자동로그인 및 더미 스크린 구현

angryPodo committed:
Merge remote-tracking branch 'origin/develop' into feat/#3-week-2

angryPodo committed:
refactor home navigation to include mypage and update ui components

angryPodo committed:
add new home screen assets and viewmodel

angryPodo committed:
Refactor: Update package structure and state management

This commit includes several refactoring changes to the package structure and state management within the application.

Key changes:

-   Moved `HomeNavigation` to `home/navigation/HomeNavigation.kt`.
-   Renamed `SignUpScreen` to `MainSignUpScreen`.
-   Deleted `SignInState.kt` and `SignUpState.kt`, replaced with model files in the specific package.
-   Created `SignInState.kt` and `SignUpState.kt` in `model` package.
-   Updated imports to reflect the new package structure.
-   Modified `HomeTopAppBar` to set tint to `Color.Unspecified`.
- Changed `AtSoptSnackbar` text color to `Black`
- Modified `SignUpPasswordScreen` to private function.
- Modified `LocalSnackBarTrigger` comment.

angryPodo committed:
feat: Rename LiveNavGraph to liveNavGraph and Update HomeImageBanner

This commit refactors the naming of `LiveNavGraph` to `liveNavGraph` for better consistency and updates `HomeImageBanner` to improve readability.

Key changes:

-   Renamed `LiveNavGraph` to `liveNavGraph` in `LiveNavigation.kt` and `MainScreen.kt`.
-   Modified `HomeImageBanner` in `HomeRoute.kt` to use `showRank` instead of `showPosition`.
-   Updated the parameter name in `HomeImageBanner` from `showPosition` to `showRank`.
- Added conditional logic to display rank based on `showRank`.
- Removed unused variable in `HomeRoute.kt`.
- Update package path to match the file location.

angryPodo committed:
feat: Enable edge-to-edge display with dark system bars

This commit enables edge-to-edge display in the app and sets the system bars (status bar and navigation bar) to dark mode.

Key changes:

-   Called `enableEdgeToEdge()` in `MainActivity`'s `onCreate()` to enable edge-to-edge display.
-   Set both `statusBarStyle` and `navigationBarStyle` to dark using `SystemBarStyle.dark(Color.Black.hashCode())`.

sohee6989 commented on app/src/main/java/org/sopt/at/core/designsystem/component/button/Image.kt:
```diff
@@ -0,0 +1,39 @@
+    modifier: Modifier = Modifier,
+    contentScale: ContentScale = ContentScale.Fit,
+    contentDescription: String? = null,
+    shape: RoundedCornerShape = CircleShape
```
shape 파라미터가 RoundedCornerShape 타입인데 기본값으로 CircleShape을 사용하신 이유가 궁금합니다.
저는 RoundedCornerShape(15.dp)처럼만 사용해봐서요..!

angryPodo replied:
먼저 RoundedCornerShape를 사용해서 원형모양을 구현했다면 컴포넌트의 크기가 커졌을때도 과연 완벽한 원형이 유지될까요?
정답은 유지되지 않습니다. 때문에 곡률을 주기 위한거라면 소희님처럼 사용하면 되지만, 원형을 만들기 위해서는 CircleShape를 사용해주는것이 안전합니다. 다른 개발자에게 해석의 리소스도 줄일 수 있겠죠?

sohee6989 commented on app/src/main/java/org/sopt/at/presentation/home/HomeRoute.kt:
```diff
@@ -0,0 +1,98 @@
+) {
+    val scrollState = rememberScrollState()
+
+    Column(
```
HomeImageBanner 만드실 때나 CategorySection 부분에서는 LazyRow를 사용하셨는데 여기서는 Column으로 하신 이유가 궁금합니다!

angryPodo replied:
먼저 리뷰어분들에게 두가지 방식을 보여드리려고 했어요 :)
또한 저는 스크롤이 과도하게 생성되지 않다면 Lazy를 사용하지 않습니다. 성능상의 차이가 미미하기 때문이에요. 때문에 현재 Home은 제한된 스크롤을 가지고 있습니다. 그래서 일반 Column에 스크롤 이벤트를 줬어요

sohee6989 commented on app/src/main/java/org/sopt/at/core/navigation/MainTabRoute.kt:
```diff
@@ -0,0 +1,3 @@
+package org.sopt.at.core.navigation
+
+interface MainTabRoute : Route
```
이후에 메인 탭 안에서 이동하는 화면들을 만든다면 MainTabRoute로 타입을 설정하는 게 좋은가요? 아니면 그냥 Route로 설정해도 괜찮나요??

angryPodo replied:
먼저 소희님도 `MainTabRoute`와 `Route`를 구분하셨는데 어떤 기준으로 나누셨나요? `MainTab`이 의도하는게 어떤건지 여쭤보고 싶어요. 그리고 `MainTabRoute`와 `Route`를 바꾼다고 해서 UI에 변화가 일어날까요?
목적지까지 이동하는 과정에서 어떤 화면과 연결되어 있는지를 생각해보시면 좋겠습니다ㅎㅎ

sayyyho commented on app/src/main/AndroidManifest.xml:
```diff
@@ -2,7 +2,12 @@
     xmlns:tools="http://schemas.android.com/tools">
 
+    <!-- 네트워크 연결 설정 추가 -->
+    <uses-permission android:name="android.permission.INTERNET" />
```
서버통신 미리 준비하시나요?

angryPodo replied:
푸하하 미리넣어놨습니다

sayyyho commented on app/src/main/java/org/sopt/at/core/designsystem/component/button/Image.kt:
```diff
@@ -0,0 +1,39 @@
+    contentScale: ContentScale = ContentScale.Fit,
+    contentDescription: String? = null,
+    shape: RoundedCornerShape = CircleShape
+) {
+    if (LocalInspectionMode.current) {
+        Image(
+            imageVector = ImageVector.vectorResource(R.drawable.ic_launcher_background),
+            contentDescription = contentDescription,
+            contentScale = contentScale,
+            modifier = modifier.clip(shape)
+        )
+    } else {
+        AsyncImage(
+            model = url,
+            contentDescription = contentDescription,
+            contentScale = contentScale,
+            modifier = modifier.clip(shape)
+        )
+    }
+}
```
이미지를 프로젝트 안에 넣지 않고 불러오는건가요?

angryPodo replied:
네! Async를 통해서 외부 이미지를 불러올때 사용하는겁니다! 지금 프로필 사진만 Coil로 불러오고 있어요! 나머진 내부!

sohee6989 reviewed (approved):
파일 관리, 화면에 padding 설정, SAA로 만드는 방법 등 여러모로 정말 많이 배워갑니다!🙇‍♀️
민재님 코드 참고해서 과제 리팩토링 열심히 해보겠습니다..!ㅎㅎ
자동 로그인 구현 해주신 부분은 더 공부해보고 궁금한거 생기면 질문 드리겠습니다.
감사합니다~~!!

sayyyho commented on app/src/main/java/org/sopt/at/core/designsystem/component/textfield/PasswordTextField.kt:
```diff
@@ -45,7 +45,7 @@ fun PasswordTextField(
         placeholder = placeholder,
         onValueChanged = { newText ->
-            if (newText.length <= maxLength && isPasswordFormat(newText)) {
+            if (newText.length <= maxLength && (newText.isEmpty() || isPasswordFormat(newText))) {
```
Good

sayyyho commented on app/src/main/java/org/sopt/at/core/designsystem/event/SnackbarTrigger.kt:
```diff
@@ -0,0 +1,7 @@
+
+import androidx.compose.runtime.staticCompositionLocalOf
+
+val LocalSnackBarTrigger = staticCompositionLocalOf<(String) -> Unit> {
+    error("이건 어떤 로직으로 동작할까요?")
+}
```
어떤 로직인지 설명 가능할까요?

angryPodo replied:
Props drilling을 찾아보시면 좋을 듯 합니다. 푸하하

sayyyho commented on app/src/main/java/org/sopt/at/core/designsystem/theme/Type.kt:
```diff
@@ -1,4 +1,4 @@
-package org.sopt.at.ui.theme
+package org.sopt.at.core.designsystem.theme
```
core.designsystem 하위에서 color,theme,type을 관리하시나요? 저는 처음봐서 배워가네요!

angryPodo replied:
아키텍처마다 다르겠지만 디자인시스템은 코어에 해당한다고 생각해서 저런 구조를 사용하는걸 선호합니다:)

sayyyho commented on app/src/main/java/org/sopt/at/core/util/extension/ModifierExt.kt:
```diff
@@ -0,0 +1,28 @@
+    this.clickable(
+        indication = null,
+        interactionSource = remember { MutableInteractionSource() },
+        onClick = { onClick() }
```
불필요하게 람다를 한 번 더 감싸고 있는 거 같은데, 람다 인스턴스 생성을 줄여 미세하게 퍼포먼스를 개선할 수 있고 () -> Unit으로 고정되어 있기 때문에, 직접 참조 방식으로 수정은 어떨까요?
```suggestion
        onClick = onClick
```

angryPodo replied:
히히 놓쳤네요 수정할게요~

angryPodo replied:
아하 저 익스텐션 함수가 inline 펑션이라서 직접참조로 바꾸게 되면 람다가 인라인되지 않고 일반 함수 객체로 전달됩니다. noinline으로 바꾸면 인라인의 성능상 이점이 없어져서요~

sayyyho commented on app/src/main/java/org/sopt/at/core/util/validation/SoptValidator.kt:
```diff
@@ -0,0 +1,14 @@
+package org.sopt.at.core.util.validation
+
+object SoptValidator {
+    private val userIdPattern = "^[a-z0-9]+$".toRegex()
+    private val passwordPattern = "^[A-Za-z0-9~!@#$%^&*()_+\\-=\\[\\]{}|;:'\",.<>/?]+$".toRegex()
+
+    fun isPasswordFormat(input: String): Boolean {
+        return input.matches(passwordPattern)
+    }
+
+    fun isUserIdFormat(input: String): Boolean {
+        return input.matches(userIdPattern)
+    }
+}
```
요긴하게 잘 쓰겠습니다....!!

sayyyho commented on app/src/main/java/org/sopt/at/data/repository/UserRepositoryImpl.kt:
```diff
@@ -0,0 +1,64 @@
+import kotlinx.coroutines.flow.map
+import org.sopt.at.domain.repository.UserRepository
+
+private val Context.dataStore: DataStore<Preferences> by preferencesDataStore(name = "user_preferences")
```
교수님 진도가 너무 빨라요....!

angryPodo replied:
그치만 1주차에 심화였는걸,,도약인가?

angryPodo committed:
feat: Enhance LazyRow item key management

This commit enhances the key management of items within LazyRow components in `CategorySection` and `HomeImageBanner`.

Key changes:

-   In `CategorySection`, `LazyRow` now uses `category` as the key for each item to prevent recomposition of already rendered items.
- In `HomeImageBanner`, `LazyRow` now uses the image resource as the key for each item.
- Updated the method to `items` and `itemsIndexed` and added the key parameter to make item position persistent and optimized.
- According to the search result, Java code review tools are important for catching the bug early, so this commit could be a good practice for the future.

JiWoo1261 reviewed (commented):
DataStore에 로그인 여부만 저장하고 있는것 같은데 굳이 DataStore을 쓰는 이유가 있을까요? 추후에 확장 가능 하도록 미리 적용해둔건지 궁금합니다.ᐟ

sayyyho commented on app/src/main/java/org/sopt/at/presentation/home/HomeRoute.kt:
```diff
@@ -0,0 +1,98 @@
+
+@Composable
+private fun HomeScreen(
+    bannerImages: ImmutableList<Int>,
+    topContent: ImmutableList<ContentList>,
```
이런 디테일함 최고네요... 저는 구현에만 급급해서 List로 했던 거 같은데 ㅠㅠㅠ

sayyyho commented on app/src/main/java/org/sopt/at/presentation/signin/SignInRoute.kt:
```diff
@@ -76,7 +109,7 @@ fun SignInScreen(
 
         Text(
-            text = "TIVING ID 로그인",
+            text = "TVING ID 로그인",
```
아 Tiving이 아니고 Tving이군요...

angryPodo replied:
ㅋㅋㅋ저도 몰랐어요..

sayyyho commented on app/src/main/java/org/sopt/at/presentation/signup/MainSignUpScreen.kt:
```diff
@@ -0,0 +1,61 @@
+    val lifecycleOwner = LocalLifecycleOwner.current
+    val snackbarTrigger = LocalSnackBarTrigger.current
+
+    LaunchedEffect(viewModel.sideEffect, lifecycleOwner) {
+        viewModel.sideEffect.flowWithLifecycle(lifecycleOwner.lifecycle).collect { effect ->
+            when (effect) {
+                is SignUpSideEffect.ShowSnackbar -> {
+                    snackbarTrigger(effect.message)
+                }
+            }
+        }
+    }
```
깔꼼..!

sayyyho commented on app/src/main/java/org/sopt/at/presentation/signup/SignUpViewModel.kt:
```diff
@@ -0,0 +1,98 @@
+    data class ShowSnackbar(val message: String) : SignUpSideEffect()
+}
+
+@HiltViewModel
+class SignUpViewModel @Inject constructor(
+    private val userRepository: UserRepository
+) : ViewModel() {
+
+    private val _state: MutableStateFlow<SignUpState> = MutableStateFlow(SignUpState())
+    val state: StateFlow<SignUpState>
+        get() = _state.asStateFlow()
+
+    private val _uiState = MutableStateFlow<UiState<Unit>>(UiState.Empty)
+    val uiState: StateFlow<UiState<Unit>>
+        get() = _uiState.asStateFlow()
```
바이블이네요. 저도 이번 리팩때 도입하겠습니다!

angryPodo replied:
```kotlin
var state: MutableStateFlow<SignUpState> = MutableStateFlow(SignUpState())
    private set
```
사실 이렇게 써도 무방하지 않나 싶긴합니다ㅎ

sayyyho reviewed (approved):
2주차 때 시험도 보시고, OB 세미나 준비까지 하시느라 정말 바쁘셨을 텐데,  
그 와중에도 코드 퀄리티를 이렇게 잘 챙겨주셔서 많이 배워갑니다!
저는 구현에만 급급해서 놓친 부분이 많았는데, 민재님 코드 보면서 정말 많이 배웠습니다👍

angryPodo commented:
>DataStore에 로그인 여부만 저장하고 있는것 같은데 굳이 DataStore을 쓰는 이유가 있을까요? 추후에 확장 가능 하도록 미리 적용해둔건지 궁금합니다.ᐟ

@JiWoo1261 음 지우님 혹시 어떤 파일을 보시고 코멘트를 주셨을까요? 데이터스토어는 로그인 여부뿐만 아니라 사용자 아이디와 비밀번호도 함께 저장하고 있습니다. 코드를 보시면 사용자 아이디와 비밀번호를 저장하고, getUserId()와 getUserPassword() 메서드도 있고, saveUserCredentials(id, password) 메서드를 통해 저장하고 있어요. DataStore는 단순히 로그인 상태뿐 아니라 자동 로그인을 위한 모든 필요 정보를 안전하게 보관하는 데 사용되고 있습니다만...ㅠ
